### PR TITLE
[Cody completion]: Improve trailing completion based on suffix string

### PR DESCRIPTION
**Pull request by vovakulikov**
Fixes https://github.com/sourcegraph/cody/issues/264

Prior to this PR, we cut off some trailing parts of completion based on the number of connected characters sequence in the begining of the suffix part (see this PR for more details https://github.com/sourcegraph/cody/pull/108). In this PR, we do simular same logic, but we try to compare strings a bit deeper based on a minimum number of edits operation over strings. If there is more than 30% of differences, we won't cut off the trailing completion. Otherwise, we cut off the completion string. This should handle cases like 

```
// Todo: cancel network request 
const abortController = new AbortController() // ← completion
const networkAbortController = new AbortController()
```

To be clear, even this comparing isn't ideal because there will be some corner cases when we will or won't cut off parts that same/different compared to the first suffix string. But it should cover more cases than the current logic. 

## Test plan
- Testing is the most tricky part here, but we need to find places where Cody generates completion similar to the suffix (next string after insertion happens). You can find a few examples of this behaviour in the issue thread https://github.com/sourcegraph/cody/issues/264

### CODE REVIEW

File: vscode/src/completions/text-processing.ts [hunk 1]
```diff
@@ -165,45 +155,46 @@ export function trimUntilSuffix(insertion: string, prefix: string, suffix: strin
...
-         * The resulting completion will be: `pnpm-store`.
-         */
-        const linePart = getFirstNCharsPreservingLeadingSpaces(line, NUMBER_OF_CHARS_TO_MATCH_AND_CUT_SUFFIX)
-        if (linePart.length && firstNonEmptySuffixLinePart.startsWith(linePart)) {
+        if (line === firstNonEmptySuffixLine) {
             insertionEnd = i
             break
         }
+
+        if (i === insertionLines.length - 1) {
```
Comment (philipp-spiess) [inline vscode/src/completions/text-processing.ts:168]: I think this is another regression, we actually do not know when to end the multi-line completion and _any_ matching line with the first non empty line will be considered the cut off. Imagine something like this:


Document:

```
function bubbleSort() {
  {$CURSOR}
	do {
   swapped = false;
        for (let i = 0; i < array.length - 1; i++) {
            if (array[i] > array[i + 1]) {
                let temp = array[i];
                array[i] = array[i + 1];
                array[i + 1] = temp;
                swapped = true;
            }
        }
   } while (swapped);
}
```

Completion:

```
let swapped;
    do {
        swapped = false;
        for (let i = 0; i < array.length - 1; i++) {
            if (array[i] > array[i + 1]) {
                let temp = array[i];
                array[i] = array[i + 1];
                array[i + 1] = temp;
                swapped = true;
            }
        }
    } while (swapped);
```

The next non-empty line is `do {` which is the second line of the completion. With your change, we'd only compare it to the last line of the completion which would man it does _not_ match, inserting the whole completion - Or am I misunderstanding?

Let's add this as a test case too if that's an issue 😬

**Review by philipp-spiess — COMMENTED** [review]
Good stuff! I left some comments inline. There are two cases where I worry we might introduce slight regressions - Will stamp once we validate if that's the case or not 👍

File: vscode/src/completions/completion.test.ts [hunk 2]
```diff
@@ -770,20 +770,54 @@ describe('Cody completions', () => {
...
-                    createCompletionResponse(`
-                    pnpm-store
-                        key: {{ runner.os }}-pnpm-{{ steps.pnpm-cache.outputs.STORE_PATH }}
-                    }`),
-                ]
-            )
+        describe('stops when the next non-empty line of the suffix matches partially', () => {
+            it('simple example', async () => {
+                const { completions } = await complete(
+                    `path: $GITHUB_WORKSPACE/vscode/.vscod-etest/${CURSOR_MARKER}
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply vscode/src/completions/completion.test.ts:776]: I think you mentioned this in our live session, but I forgot to update this case. It made sense to me when you told me, but I agree it's difficult to understand if you don't know how indention is handled in the trim logic. Maybe `\n` will improve mental mode here.

File: vscode/src/completions/text-processing.ts [hunk 3]
```diff
@@ -140,23 +135,18 @@ const NUMBER_OF_CHARS_TO_MATCH_AND_CUT_SUFFIX = 15
...
-        if (line.trim().length > 0) {
-            firstNonEmptySuffixLine = line
-            break
-        }
-    }
+
+    const firstNonEmptySuffixLine = getFirstNonEmptyLine(suffix)
+
+    // TODO: Handle case for inline suffix - remove same trailing sequence from insertion
+    // if we already have the same sequence in suffix
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply vscode/src/completions/text-processing.ts:142]: Yeah, maybe it's irrelevant anymore since you fixed it later in the insertion pipeline, but I thought if we do post-procession of the insertion string, we should do it in one place. (here or in multiline trimming) 

I meant this case 
`console.log('Hello ${cursor}')` → `console.log('Hello world ')')` where `world ')` is completion. I know that this won't work at the moment in this case because we replace it later on vs code side. I thought we should handle it in the text processing and remove trailing completion if we already have the same sequence in the suffix.

But even here, it won't be ideal because merging isn't that straightforward when you handle multiline completions. What if we have this example with bubbleSort and arguments 

```
function bubbleSort(${cursor}) {
   while (...) {...} <- completion 
}
```

in this case completion is 
```
}) {
 while (...) {...}
}
```
But we already have the`)` suffix in the first string, so we should delete it from the prefix, and we can't do it from text processing. 

Sorry for the long comment here. I will remove this comment. Or maybe will add a separate one about special cases like this and that they are handled on a different level.

File: vscode/src/completions/text-processing.ts [hunk 4]
```diff
@@ -165,45 +155,46 @@ export function trimUntilSuffix(insertion: string, prefix: string, suffix: strin
         if (i === 0) {
             const lastNewlineOfPrefix = prefix.lastIndexOf('\n')
             line = prefix.slice(lastNewlineOfPrefix + 1) + line
+
+            // Skip any procession for the first (inline string insertion/completion)
+            continue
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply vscode/src/completions/text-processing.ts:160]: Oh, You're right. About `console.log(2)`, I think it's all about corner cases, we can add some special logic about it that we exclude `console.log()` call from compare algorithm and compare only its arguments (in this case 'two' and `three`) and this would give us a proper completion. I would do it in a separate PR though, I think there are a lot of other cases when checks like this may make some difference.

File: vscode/src/completions/text-processing.ts (hunk 1 above)
Comment (vovakulikov) [reply vscode/src/completions/text-processing.ts:168]: I see your point here, I checked it on the main, and it doesn't seem like a regression, though, since the main also fails here. 

Screenshot 2023-07-21 at 13 37 23

The problem here is that we always compare the current line with the first suffix line. It should be like if the current line matches with the current line in the suffix (starting with the first non-empty suffix line) we should skip and next time compare the current completion line with the current suffix line), if we want to handle cases with multi-block lines like this, we need to float in the suffix line and don't compare only the first non-empty line. 

I will try to fix it but it seems like a big change that can break other cases, if it's big enough I suggest adding this in the separate PR. I'll keep you posted

File: vscode/src/completions/text-processing.ts [hunk 5]
```diff
@@ -165,45 +155,46 @@ export function trimUntilSuffix(insertion: string, prefix: string, suffix: strin
...
-         */
-        const linePart = getFirstNCharsPreservingLeadingSpaces(line, NUMBER_OF_CHARS_TO_MATCH_AND_CUT_SUFFIX)
-        if (linePart.length && firstNonEmptySuffixLinePart.startsWith(linePart)) {
+        if (line === firstNonEmptySuffixLine) {
             insertionEnd = i
             break
         }
+
+        if (i === insertionLines.length - 1) {
+            if (isAlmostTheSameString(line.trim(), firstNonEmptySuffixLine.trim())) {
```
[the remark this answers is not in the document]
Comment (vovakulikov) [reply vscode/src/completions/text-processing.ts:169]: True, I probably still want to have trim because fewer symbols mean we have a smaller edit distance matrix, I will add an indentation check there

File: vscode/src/completions/text-processing.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (philipp-spiess) [reply vscode/src/completions/text-processing.ts:142]: Thanks for the comment, I understand what you mean. I am really curious how VS Code does this right now. Maybe they just pick the first occurence of the same line suffix in the completion and do a character based diffing 🤔

File: vscode/src/completions/text-processing.ts (hunk 1 above)
Comment (philipp-spiess) [reply vscode/src/completions/text-processing.ts]: @vovakulikov 

Hm I think the indentation in your test case is a bit off. I think I have added indentation errors when I was explaining the case, sorry about that!

> The problem here is that we always compare the current line with the first suffix line.

My understand is this is what we did _before_ but with your change here, especially this line `if (i === insertionLines.length - 1) {` we will now only compare the next non empty-line with the last line of the completion.

Comment (vovakulikov) [reply vscode/src/completions/text-processing.ts]: @philipp-spiess I just ran this test on this branch PR, and it also passed since my last fix about indentations and first-line handling. 

But I still believe that the current logic isn't sufficient. Either main or this PR, we make decisions about suffixes only by its one non-empty line (in cases like this where we have to compare block by block instead of line by line, we should be smarter, I think), I will try to improve this in the follow-up PR.

Comment (philipp-spiess) [reply vscode/src/completions/text-processing.ts]: Oh yeah I see that you removed this condition in your latest version. Makes sense!

File: vscode/src/completions/text-processing.ts [hunk 6]
```diff
@@ -167,20 +157,9 @@ export function trimUntilSuffix(insertion: string, prefix: string, suffix: strin
...
-         *
-         * And the completions is:
-         * `pnpm-store\n  key: ${{ runner.os }}-pnpm-${{ steps.pnpm-cache.outputs.STORE_PATH }}`
-         *
-         * We cut the completion on the `  key:` part to avoid duplicating the suffix.
-         * The resulting completion will be: `pnpm-store`.
-         */
-        const linePart = getFirstNCharsPreservingLeadingSpaces(line, NUMBER_OF_CHARS_TO_MATCH_AND_CUT_SUFFIX)
-        if (linePart.length && firstNonEmptySuffixLinePart.startsWith(linePart)) {
+        const isSameIndentation = indentation(line) === indentation(firstNonEmptySuffixLine)
```
Comment (philipp-spiess) [inline vscode/src/completions/text-processing.ts:160]: Only last nit: Should this be 

```suggestion
        const isSameIndentation = indentation(line) <= indentation(firstNonEmptySuffixLine)
```

just to make sure we have a bit more resilience against screwed up indentation haha

**Review by philipp-spiess — APPROVED** [review]
Nice! Thank you so much for plowing through all the test cases etc. I think this is good to go! Will be cool to see it's impact in our top line metric 😎

**Comment by vovakulikov** [comment]
Ok, it looks like everything is ready to merge, I hope this will work as we think it works =)
